validator/sawtooth_validator/execution/scheduler_parallel.py

```diff
@@ -111,14 +105,44 @@ def get_ancestors(self, address):
             node = node.children[token]
             yield node
 
+
+class Predecessors:
+    def __init__(self):
+        self.readers = []
+        self.writer = None
+
+    def __repr__(self):
+        return repr({
+            'readers': self.readers,
+            'writer': self.writer,
+        })
+
+
+class PredecessorTree:
+    def __init__(self, token_size=2):
+        self._tree = Tree(token_size=token_size)
+
+    def __repr__(self):
+        return repr(self._tree)
+
+    def _insert(self, address):
+        node = self._tree.insert(address)
+
+        if node.data is None:
+            node.data = Predecessors()
```

Comment:
Can you just pass the new instance of Predecessor to `self._tree.insert`?

Reply:
No. The logic of `insert` (maybe it needs a different name) is to get the node at address if it exists, and create it if it doesn't. In the former case, we don't necessarily want to overwrite the existing data.